Added the component specification

sarahmaddox opened the pull request:
fixes https://github.com/kubeflow/website/issues/570

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/website/604)
<!-- Reviewable:end -->

sarahmaddox commented:
Preview: https://deploy-preview-604--competent-brattain-de2d6d.netlify.com/docs/pipelines/reference/component-spec/

sarahmaddox commented:
/assign @Ark-kun 
/assign @abhi-g 

Note: I'll link this page to the component development guide https://github.com/kubeflow/website/pull/601 (and vice versa) once both PRs are merged.

paveldournov commented on content/docs/pipelines/reference/component-spec.md:
```diff
@@ -0,0 +1,135 @@
+replaced by a path. (The path can possibly point to a mounted output volume.)
+
+* In `component.yaml': `{outputPath: Trained model}`
+* Resulting Argo YAML: `"/outputs/trained_model/data"`
```
is trained_model directory constructed from the "Trained model" name? In the docs - can we always keep the names matching? Or we need to reference a clear translation logic here.

sarahmaddox replied:
@Ark-kun

Ark-kun replied:
>is trained_model directory constructed from the "Trained model" name?

Yes.

> can we always keep the names matching?

We'll fix this part to make the names match. This is only a path though and the user does not interact with it.

> Or we need to reference a clear translation logic here.

We can describe the translation logic. But half of the purpose of generating the paths (as opposed to hard-coding the paths) is that it's opaque and can be changed later. We could as well just generate UUID paths like `/outputs/5c9b2d651e6c01cb41d834faa77925157d5d1599/data`. The user should not really care about or depend on some concrete logic.

Ark-kun replied:
Maybe we should not mention Argo here at all.

Ark-kun replied:
I've added comments with replacement text. We should not mention Argo. (Remove this line)

sarahmaddox replied:
Done.

paveldournov commented on content/docs/pipelines/reference/component-spec.md:
```diff
@@ -0,0 +1,135 @@
+    * `type`: Specifies the type of input/output. The types are used
+        as hints for pipeline authors and can be used by the pipeline system/UI
+        to validate arguments and connections between components. Basic types
+        are **String**, **Integer**, **Number**, **Boolean**.
```
Is this a full list of supported types? If not - can we have a reference here to the full list?

sarahmaddox replied:
Done.

sarahmaddox commented:
/unassign @abhi-g 
/assign @paveldournov

Ark-kun commented on content/docs/pipelines/reference/component-spec.md:
```diff
@@ -0,0 +1,137 @@
+
+The placeholder is replaced by the contents of the input argument.
+
+* In `component.yaml`: `{inputValue: Rounds}`
```
In `component.yaml`:
```command: [program.py, --rounds, {inputValue: Rounds}]```
In pipeline code:
```task1 = component1(rounds=150)```
Resulting command-line (the placeholder is substituted):
```program.py --rounds 150```

sarahmaddox replied:
Thanks. Done.

Ark-kun commented on content/docs/pipelines/reference/component-spec.md:
```diff
@@ -0,0 +1,137 @@
+Output paths are filled in by the pipeline system. The Output construct is
+replaced by a path. (The path can possibly point to a mounted output volume.)
+
+* In `component.yaml`: `{outputPath: Trained model}`
```
In `component.yaml`:
```command: [program.py, --out-model, {outputPath: trained_model}]```
In pipeline code:
```task1 = component1()```
Resulting command-line (the placeholder is substituted with generated path):
```program.py --out-model /outputs/trained_model/data```

sarahmaddox replied:
Done.

Ark-kun commented on content/docs/pipelines/reference/component-spec.md:
```diff
@@ -0,0 +1,137 @@
+
+### Outputs
+
+Output paths are filled in by the pipeline system. The Output construct is
```
```suggestion
Output paths are filled in by the pipeline system. The outputPath placeholder is
```

sarahmaddox replied:
Done.

Ark-kun commented on content/docs/pipelines/reference/component-spec.md:
```diff
@@ -0,0 +1,137 @@
+
+Output paths are filled in by the pipeline system. The Output construct is
+replaced by a path. (The path can possibly point to a mounted output volume.)
+
```
```suggestion
The parent directories of the path might or might not not exist.
The program must handle both cases without errors.

```

sarahmaddox replied:
Done.

Ark-kun commented on content/docs/pipelines/reference/component-spec.md:
```diff
@@ -0,0 +1,137 @@
+The placeholder is replaced by the contents of the input argument.
+
+* In `component.yaml`: `{inputValue: Rounds}`
+* Resulting Argo YAML: `"{{input.parameters.rounds}}"`
```
```suggestion
```

sarahmaddox replied:
Done.

sarahmaddox commented:
Thanks for the review. :) I've addressed all the comments. PTAL https://deploy-preview-604--competent-brattain-de2d6d.netlify.com/docs/pipelines/reference/component-spec/

Ark-kun commented:
/lgtm
/approve
